pkg/reconciler/native.go

```diff
@@ -286,23 +347,55 @@ func (rec *NativeReconciler) purge(excluded map[string]bool, componentId string)
 			if excluded[resourceID] {
 				continue
 			}
-			if objectMeta.GetAnnotations() != nil && objectMeta.GetAnnotations()[BanzaiCloudManagedComponent] == componentId {
+			if objectMeta.GetAnnotations() != nil && objectMeta.GetAnnotations()[types.BanzaiCloudManagedComponent] == componentId {
 				rec.Log.Info("pruning unmmanaged resource",
 					"name", objectMeta.GetName(),
 					"namespace", objectMeta.GetNamespace(),
 					"group", gvk.Group,
 					"version", gvk.Version,
 					"listKind", gvk.Kind)
-				if err := rec.Client.Delete(context.TODO(), &o); err != nil {
+				if err := rec.Client.Delete(context.TODO(), &o); err != nil && !k8serrors.IsNotFound(err) {
 					allErr = errors.Combine(allErr, err)
+				} else {
+					rec.addReconciledObjectState(ReconciledObjectStatePurged, o.DeepCopy())
 				}
 			}
 		}
 	}
 	return allErr
 }
 
-func (rec *NativeReconciler) addAnnotation(o runtime.Object, componentId string) (metav1.Object, error) {
+type reconciledObjectState string
+
+const (
+	ReconciledObjectStateAbsent  reconciledObjectState = "Absent"
+	ReconciledObjectStatePresent reconciledObjectState = "Present"
+	ReconciledObjectStatePurged  reconciledObjectState = "Purged"
+)
+
+func (rec *NativeReconciler) initReconciledObjectStates() {
+	rec.reconciledObjectStates = map[reconciledObjectState][]runtime.Object{
+		ReconciledObjectStateAbsent:  make([]runtime.Object, 0),
+		ReconciledObjectStatePresent: make([]runtime.Object, 0),
+		ReconciledObjectStatePurged:  make([]runtime.Object, 0),
+	}
+}
+
+func (rec *NativeReconciler) addReconciledObjectState(state reconciledObjectState, o runtime.Object) {
+	rec.reconciledObjectStates[state] = append(rec.reconciledObjectStates[state], o)
+}
+
+func (rec *NativeReconciler) GetReconciledObjectWithState(state reconciledObjectState) []runtime.Object {
+	return rec.reconciledObjectStates[state]
+}
+
+func (rec *NativeReconciler) addRelatedToAnnotation(objectMeta, ownerMeta metav1.Object) {
+	annotations := objectMeta.GetAnnotations()
+	annotations[types.BanzaiCloudRelatedTo] = utils.ObjectKeyFromObjectMeta(ownerMeta).String()
```

Comment:
`annotations` might be `nil` which is bad when trying to set values on a map
```suggestion
	if annotations == nil {
		annotations = make(map[string]string)
	}
	annotations[types.BanzaiCloudRelatedTo] = utils.ObjectKeyFromObjectMeta(ownerMeta).String()
```